sensors: check battery array bounds

jlecoeur opened the pull request:
@dagar fixes CI in #7644 

When trying to move the selected power source to the first publication instance on systems where there is only one power source, the compiler issued the warning
```
../src/modules/sensors/sensors.cpp:512:54: error: array subscript is outside array bounds [-Werror=array-bounds]
         tmp_h = _battery_pub[_battery_pub_intance0ndx];
```
because it could not verify that _battery_pub_intance0ndx would always be 0.

Fix: 
Wrap the block between `#if BOARD_NUMBER_BRICKS > 1 [...] #endif` to ensure no out of bound subscript and to remove the warning.

TODO:
- [ ] Test on V4 hardware according to @davids5 's test plan (see below)

LorenzMeier commented:
@davids5 Can you review?

davids5 commented on src/modules/sensors/sensors.cpp:
```diff
@@ -167,12 +167,12 @@ class Sensors : public ModuleBase<Sensors>
 	int 		_params_sub;			/**< notification of parameter updates */
 
-	orb_advert_t	_sensor_pub;			/**< combined sensor data topic */
-	orb_advert_t	_battery_pub[BOARD_NUMBER_BRICKS] = {nullptr};			/**< battery status */
```
@dagar are we still preferring to this style init?

dagar replied:
I strongly prefer to have the initialization in the definition. It makes it much easier to see that everything is covered and avoids the initialization order problems.

I also have a slight preference for c++11 uniform initialization which has some advantages for non primitive data types. https://www.youtube.com/watch?v=yKYcybUEX3I&feature=youtu.be&t=23m

All that is kind of irrelevant though because we don't have written coding standards, so for now I'd say consistency wins. Doing a pass to make sure everything is properly initialized in the style we prefer could make for good introductory CQWG tasks.

jlecoeur replied:
+1 let's keep this change for consistency, then I'm happy to help uniformize the init style.

davids5 commented on src/modules/sensors/sensors.cpp:
```diff
@@ -507,7 +509,7 @@ Sensors::adc_poll(struct sensor_combined_s &raw)
 
 							/* Move the selected_source to instance 0 */
-							if (_battery_pub_intance0ndx != b) {
+							if (BOARD_NUMBER_BRICKS > 1 && _battery_pub_intance0ndx != b) {
```
Was there a compiler warning?

When BOARD_NUMBER_BRICKS == 1, b can only be 0.  The initial condition is _battery_pub_intance0ndx == 0
So  _battery_pub_intance0ndx != b is always false.

jlecoeur replied:
@davids5 yes sorry for the lack of explanation. 
This is a fix for a compiler warning in the gcc7 CI build newly added by @dagar in #7644
```
../src/modules/sensors/sensors.cpp:512:54: error: array subscript is outside array bounds [-Werror=array-bounds]
         tmp_h = _battery_pub[_battery_pub_intance0ndx];
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
compilation terminated due to -Wfatal-errors.
cc1plus: all warnings being treated as errors
[170/535] Building CXX object src/modu.../modules__sensors.dir/parameters.cpp.o
ninja: build stopped: subcommand failed.
Makefile:158: recipe for target 'posix_sitl_default' failed
```
The full build log is here: https://travis-ci.org/PX4/Firmware/jobs/255098609

I think the compiler is trying to be too smart. If this block is executed, then _battery_pub_instance0ndx will be out the f bounds. 
Because _battery_pub_instance0ndx is initialized to 0 and never modified elsewhere the block cannot be executed, but apparently the compiler cannot see that.

I tried to move the initialization of the variable to the cpp file, but that did not help the compiler. I left it there because this is done like this elsewhere, but I can revert if you prefer.

davids5 replied:
@jlecoeur - Thank you for the details. 

The irony is that once #7618 comes in. We may need to refactor this or make the retrieval  # bricks, and details a runtime decision. This is need so that on V3 HW it looks like dual brick (not Servo power).  But on V2 it would return 1 and indicate (Servo power) , V3 would return 2.

We need to insure that `selected_source`  [is set](https://github.com/jlecoeur/Firmware/blob/7a061ebab9ef383fc33b8c0d8e4c198060238065/src/modules/sensors/sensors.cpp#L547), and there is lazy initialization on the pointers so it is passive, until the 
subscription pointers are set. 

Since the fix is effectively an #if BOARD_NUMBER_BRICKS > 1, should we wrap the block that way and add a comment? 

 If we change it to with #if BOARD_NUMBER_BRICKS == 1 and 
```
#if BOARD_NUMBER_BRICKS == 1 
   selected_source  = 0;
#else 
Code as is
#ifdef 
```
>I tried to move the initialization of the variable to the cpp file, but that did not help the compiler. I left it there because this is done like this elsewhere, but I can revert if you prefer.

Let's let @dagar chime in on that one.

davids5 commented:
Test plan:

IIRC: but it will make sense; 

For backward compatibility, the highest priority (lowest numbered) brick chosen by the HW power controller should always be in instance 0.

1) V4Pro boot with 1 brick on Power 1. No USB
verify 
`listener battery_status 1 0`
```
TOPIC: battery_status instance 0 #1
xxxx
connected: True
system_source: True
priority: 0
warning: 0
```
`listener battery_status 1 1`
```
TOPIC: battery_status instance 1 #1
connected: False
system_source: False
priority: 0
warning: 0
```
and  
`listener system_power'
```
TOPIC: system_power instance 0 #1
timestamp: 5882493
voltage5V_v:   4.8968
voltage3V3_v:   0.0000
v3v3_valid: 0
usb_connected: 0
brick_valid: 1
usb_vaild: 0
servo_valid: 0
periph_5V_OC: 0
hipower_5V_OC: 0
```
2) V4Pro boot with 1 brick on Power 2. No USB
verify 
`listener battery_status 1 0`
```
TOPIC: battery_status instance 0 #1
xxxx
connected: True
system_source: True
priority: 1
warning: 0
```
`listener battery_status 1 1`
```
TOPIC: battery_status instance 1 #1
connected: False
system_source: False
priority: 0
warning: 0
```
and  
`listener system_power'
```
TOPIC: system_power instance 0 #1
timestamp: 5882493
voltage5V_v:   4.8968
voltage3V3_v:   0.0000
v3v3_valid: 0
usb_connected: 0
brick_valid: 2
usb_vaild: 0
servo_valid: 0
periph_5V_OC: 0
hipower_5V_OC: 0
```

3) V4Pro boot with 2 bricks on Power 1&2. (use usb too)
Then move them around....
| B1 | B2 |
|----|----|
| 0  | 0  | 
| 1  | 0  | 
| 0  | 1  | 
| 1  | 1  | 

verify 
`listener battery_status 1 0`
```
TOPIC: battery_status instance 0 #1
xxxx
connected: True When a brick is connected
system_source: True When a brick is connected
priority: Brick number-1 = 0 for Brick 1 Only, 1 for Brick 2 Only, 0 with both
warning: x
```
`listener battery_status 1 1`
```
TOPIC: battery_status instance 1 #1
connected: True When a brick is connected
system_source: False
priority: 1 when 1 & 2 connected
warning: x
```
and  
`listener system_power'
```
TOPIC: system_power instance 0 #1
timestamp: 5882493
voltage5V_v:   4.8968
voltage3V3_v:   0.0000
v3v3_valid: 0
usb_connected: Follows USB connected
brick_valid: 0|1|2|3 None, Brick 1, Brick 2, both
usb_vaild: Follows USB connected
servo_valid: 0
periph_5V_OC: 0
hipower_5V_OC: 0
```

jlecoeur commented:
@davids5 I did these changes and rebased:
- reverted the initialization style
- used `#if BOARD_NUMBER_BRICKS > 1` to try to swap instances only when possible
- ensure `selected_source` is set (to 0 if no power source is valid) *is this the wanted behaviour?*
- moved the instance swapping block outside the for loop to be able to check whether `selected_source<0` after all channels were tested

I do not have V4 hardware to test.

davids5 commented:
@jlecoeur The clean up is good. (orb_advert_t tmp_h) but these changes, change the functionality of the code in way that will give the wrong results.

We have to have a valid brick to set the `selected_source`. The pointer shuffling and tracking which slot is instance 0 (_battery_pub_intance0ndx) can only happen once the subscriptions are created.

> ensure selected_source is set (to 0 if no power source is valid) is this the wanted behaviour?

No. 'selected_source' should be -1 until the subscriptions are created in the `orb_publish_auto(ORB_ID(battery_status), ` That is where the lazy initialization happens. and it may stay -1 if the power source is not a brick (i.e. USB)

jlecoeur commented:
> We need to insure that selected_source is set, and there is lazy initialization on the pointers so it is passive, until the subscription pointers are set.

Ok I misunderstood that comment.

I should be better now @davids5

jlecoeur commented:
@PX4TestFlights here is the [V4Pro binary](https://github.com/PX4/Firmware/files/1180332/px4fmu-v4pro_default.zip) to test on the bench with all combination of plugged/unplugged power bricks according to the plan above, thanks!

davids5 commented:
@jlecoeur looks like there is a CI system fail - `Partial System Outage - components affected: Heroku` Hopefully that will clear up, and we can proceed with merge.

@PX4TestFlights - please also build and test on FMUv2.

CarlOlsson commented:
On current master I can't run 
```make posix_sitl_default gazebo_standard_vtol```
because I get the error
```
/home/carl/repo/Standard_PX4/Firmware/src/modules/sensors/sensors.cpp:512:54: error: array subscript is outside array bounds [-Werror=array-bounds]
         tmp_h = _battery_pub[_battery_pub_intance0ndx];
```
However this PR fixes it

LorenzMeier commented:
@PX4TestFlights Any updates on flight tests?

Avysuarez commented:
Test with Pixhawk 3 PRO (V4) 
1 brick on Power 1
![1 brick](https://user-images.githubusercontent.com/12182176/28849086-4ce27e64-76ca-11e7-9c64-1837c08f2984.jpg)
1 brick on power 2
![3](https://user-images.githubusercontent.com/12182176/28849107-65664740-76ca-11e7-8a33-6b0751db864e.jpg)
2 bricks on Power 1 and 2
![5](https://user-images.githubusercontent.com/12182176/28849114-7db303c4-76ca-11e7-9b35-24a1cd017e7c.jpg)

davids5 commented:
@Avysuarez 

There are 3 more tests:

V4Pro boot with 2 bricks on Power 1&2. Use usb too.
Then move them around....
| B1 | B2 |
|----|----|
| 0 | 0 | - No Bricks.
| 1 | 0 |  only Brick 1
| 0 | 1 | only Brick 2
| 1 | 1 | - Done

davids5 commented:
@LorenzMeier once the last 3 test a results are posted - I will review and will merge.

Avysuarez commented:
@davids5 
No bricks
![no bricks](https://user-images.githubusercontent.com/12182176/28850502-d235f8a0-76d1-11e7-8592-4863761b5259.jpg)
Only brick 1
![brick1](https://user-images.githubusercontent.com/12182176/28850507-e0939dee-76d1-11e7-95f4-7e173acadb26.jpg)
Only brick 2
![brick2](https://user-images.githubusercontent.com/12182176/28850516-ef7b773c-76d1-11e7-94c4-78a5c105e26a.jpg)

davids5 commented:
@Avysuarez - Thank you!

jlecoeur commented:
Thanks @davids5 @Avysuarez !
